Add Device Attributes (#98)
memfault/memfault-docs: docs/platform/device-attributes.mdx

---
title: Device Attributes
---

> You can store custom attributes for each of your devices devices.

Reviewer: ```suggestion
You can store custom attributes for each of your devices.
```

:::info Attributes and Metrics

Any metric that a device reports through the Metric SDK automatically becomes a
device attribute (the value being the latest reading). Therefore metric and

> attribute definitions are one and the same.

Reviewer: One and the same means exactly the same -- but the terms are not synonymous. A device's attributes are the set of most recently read metrics?

Reviewer 2: I'd weaken this as we are planning to extend to settings per metric if they are an attribute in the first place.

> You can configure device attributes through the metric settings.

Author: @HBehrens are thinking about moving towards "Attributes" as the general term, and an attribute can be flagged to also be a "device metric" (which by default they all are). Will finetune this in a later iteration.

:::

## Defining an Attribute

An attribute is defined the first time you write it. You can view all defined

> properties going to **Project Settings** and then **Metrics**. An attribute

Reviewer: It feels surprising to me that you define and edit attributes in a setting called "metrics" -- maybe we should rename that to attributes, or otherwise change these docs to say this is how you define metrics (not attributes)?

It's small but just calling out the inconsistency

Reviewer 2: Yes, I keep calling this out towards @chrisc11 – we need an umbrella term that describes the structure of "aggregated readings", "readings, not stored in TSDB (future)", "last-seen values per device aka attributes"

definition can have a few properties:

- A `string_key` which devices and APIs can use to identify the attribute.
- Whether the data type is an `INT` (default), `FLOAT`, `BOOL`, or `STRING`.
- Whether or not the value is `updatable` (or write-once).
- Whether it is a `favorite` attribute, which will cause it to appear more
  prominently in a few places in the web app.
- The value `range` which informs charts and other components.

> - A set of `ignored_values` (like `-1`) which should be ignored.

Reviewer 2: Aren't all of these values equally relevant to metrics in general? Do we have a place to put this instead so we can link from here?

Author: As we consider making attributes the umbrella term, we would instead link here.

![Example of some attribute definitions](/img/blog/2021-08-17-device-properties-settings.png)

## Viewing an Attribute

Find the device you want to view the attributes for (for example by browsing
your devices on the **Devices** page) and select the **Properties** tab. You'll

> see all the properties known for this project and the respective value for the

Reviewer 2: Is this outdated? I thought we'd called this feature (incl. the tab) "Attributes"?

device you're currently looking at.

> ## Writing an Attribute

Reviewer 2: "Setting"?

Author: ```suggestion
## Setting an Attribute
```

There are three ways to write an attribute.

1. using the [Metrics API](/docs/mcu/metrics-api),
2. using the
   [REST API](https://api-docs.memfault.com/#f4e2584a-36f2-4184-82ec-3b9d0108588c),
   or
3. using the web application.

:::caution

> Because devices can update attributes, anyone who holds your non-secret

Reviewer: ```suggestion
Because devices can update attributes, anyone who holds your 
```

You are of course technically correct, but I'm wondering if that's still something we want to call out here using that language

Author: It's pretty important customers understand this and that they can not trust attributes being correct/protected from bad actors.

**Project Key** can write or overwrite these attributes.

:::

### Using the Web Application

Visit the respective device page as described in
[Viewing an Attribute](#viewing-an-attribute). Click on the _pencil icon_ next
to the value, enter the new value, and finally **Save** it.

![Example of editing an attribute via the web application](/img/blog/2021-08-17-device-properties-edit.png)

## Searching by Attribute

You can search devices by attributes. Visiting the **Devices** page, open the
**Filter** panel and select the **Custom Metric** you want to search by.

![Example of filtering by an attribute](/img/blog/2021-08-17-device-properties-filter.png)

> You can search by multiple attributes using the **+** icon.

Reviewer 2: consistency: above we said _pencil icon_ (italic) here, it's **+** (bold)

I wonder if we should find a way to re-use the actual icons from our app in the docs…

Author: ```suggestion
You can search by multiple attributes using the _plus icon_.
```

:::note Searching by Attribute vs Metric Readings

If you select a **During** time frame, devices will be searched by their metric

> readings during that time frame, not by their device attributes.

Reviewer: Maybe just re-iterate the difference?
```suggestion
readings during that time frame, not by their device attributes (the most recently reported metric).
```

---
title: Device Attributes
---

You can store custom attributes for each of your devices devices.

:::info Attributes and Metrics

> Any metric that a device reports through the Metric SDK automatically becomes a
> device attribute (the value being the latest reading). Therefore metric and
> attribute definitions are one and the same.

Reviewer 3: ```suggestion
Any metric that a device reports through the Metric SDK automatically populates a device attribute with its latest reading. For example, sending `battery_pct` as a metric will populate a `battery_pct` attribute on the device with the last known battery level read from the device. As such, you can configure both metrics and attributes via the Metrics section of the Settings page.
```

:::

## Defining an Attribute

An attribute is defined the first time you write it. You can view all defined
properties going to **Project Settings** and then **Metrics**. An attribute
definition can have a few properties:

- A `string_key` which devices and APIs can use to identify the attribute.
- Whether the data type is an `INT` (default), `FLOAT`, `BOOL`, or `STRING`.
- Whether or not the value is `updatable` (or write-once).
- Whether it is a `favorite` attribute, which will cause it to appear more
  prominently in a few places in the web app.
- The value `range` which informs charts and other components.
- A set of `ignored_values` (like `-1`) which should be ignored.

![Example of some attribute definitions](/img/blog/2021-08-17-device-properties-settings.png)

## Viewing an Attribute

Find the device you want to view the attributes for (for example by browsing

> your devices on the **Devices** page) and select the **Properties** tab. You'll

Author: ```suggestion
your devices on the **Devices** page) and select the **Attributes** tab. You'll
```

1. using the [Metrics API](/docs/mcu/metrics-api),
2. using the
   [REST API](https://api-docs.memfault.com/#f4e2584a-36f2-4184-82ec-3b9d0108588c),
   or
3. using the web application.

:::caution

Because devices can update attributes, anyone who holds your non-secret
**Project Key** can write or overwrite these attributes.

:::

### Using the Web Application

Visit the respective device page as described in
[Viewing an Attribute](#viewing-an-attribute). Click on the _pencil icon_ next
to the value, enter the new value, and finally **Save** it.

![Example of editing an attribute via the web application](/img/blog/2021-08-17-device-properties-edit.png)

## Searching by Attribute

You can search devices by attributes. Visiting the **Devices** page, open the
**Filter** panel and select the **Custom Metric** you want to search by.

![Example of filtering by an attribute](/img/blog/2021-08-17-device-properties-filter.png)

You can search by multiple attributes using the **+** icon.

:::note Searching by Attribute vs Metric Readings

If you select a **During** time frame, devices will be searched by their metric
readings during that time frame, not by their device attributes (the most

> recently reported metric).

Reviewer 2: "latest reported value"?

Author: ```suggestion
latest reported metric).
```
